maven-core/src/main/java/org/apache/maven/execution/DefaultBuildResumptionManager.java

```diff
@@ -0,0 +1,308 @@
+    }
+
+    /**
+     * Projects after the first failed project could have succeeded by using -T or --fail-at-end.
+     * These projects can be skipped from later builds.
+     * This is not the case these projects are dependent on one of the failed projects.
+     * @param result The result of the Maven build.
+     * @param failedProjects The list of failed projects in the build.
+     * @param resumeFromProject The project where the build will be resumed with in the next run.
+     * @return An optional containing a comma separated list of projects which can be skipped,
+     *   or an empty optional if no projects can be skipped.
+     */
+    private Optional<String> determineProjectsToSkip( MavenExecutionResult result, List<MavenProject> failedProjects,
+                                                      MavenProject resumeFromProject )
+    {
+        List<MavenProject> allProjects = result.getTopologicallySortedProjects();
+        int resumeFromProjectIndex = allProjects.indexOf( resumeFromProject );
+        List<MavenProject> remainingProjects = allProjects.subList( resumeFromProjectIndex + 1, allProjects.size() );
+
+        List<GroupArtifactPair> failedProjectsGAList = failedProjects.stream()
+                .map( GroupArtifactPair::new )
+                .collect( Collectors.toList() );
+
+        String projectsToSkip = remainingProjects.stream()
+                .filter( project -> result.getBuildSummary( project ) instanceof BuildSuccess )
+                .filter( project -> hasNoDependencyOnProjects( project, failedProjectsGAList ) )
+                .map( project -> String.format( "%s:%s", project.getGroupId(), project.getArtifactId() ) )
+                .collect( Collectors.joining( PROPERTY_DELIMITER ) );
+
+        if ( !StringUtils.isEmpty( projectsToSkip ) )
+        {
+            return Optional.of( projectsToSkip );
+        }
+
+        return Optional.empty();
+    }
+
+    private boolean hasNoDependencyOnProjects( MavenProject project, List<GroupArtifactPair> projectsGAs )
+    {
+        return project.getDependencies().stream()
+                .map( GroupArtifactPair::new )
+                .noneMatch( projectsGAs::contains );
+    }
+
+    private boolean writeResumptionFile( MavenProject rootProject, Properties properties )
+    {
+        Path resumeProperties = Paths.get( rootProject.getBuild().getDirectory(), RESUME_PROPERTIES_FILENAME );
+        try
+        {
+            Files.createDirectories( resumeProperties.getParent() );
+            try ( Writer writer = Files.newBufferedWriter( resumeProperties ) )
+            {
+                properties.store( writer, null );
+            }
+        }
+        catch ( IOException e )
+        {
+            logger.warn( "Could not create " + RESUME_PROPERTIES_FILENAME + " file. ", e );
+            return false;
+        }
+
+        return true;
+    }
+
+    private Properties loadResumptionFile( String rootBuildDirectory )
```

Comment:
The signature is inconsistent with `writeResumptionFile()`. once you pass the string directly, on the other one you pass the Maven project. Any reason for this?

Reply:
I agree, this would be nicer when they are consistent. Thanks!

Reply:
In general, if you know a String is actually something else (like a `Path`), please pass it as the dedicated instance.